### ISSUE TITLE
Video-to-Video model finetuning

### ISSUE THREAD

**Issue by hpoghos** [modelscope/modelscope#643]
Hi, I'm trying to finetune the `damo/Video-to-Video` model.  But after a couple of iteration my results degenerate to noise.
Could you please take a look and see if the following alignes with your training process.

I'm initializing the `unet`, `vae` and `text encoder` unsing the official weights.
```
model_ms = VideoToVideo(model_dir="modules/Video-to-Video")
vae = model_ms.autoencoder
unet = model_ms.generator
text_encoder = model_ms.clip_encoder
```
and use the following training scheduler, the (sigmas are calculated from betas obtained using [cosine beta schedule](https://github.com/modelscope/modelscope/issues/606)).
```
class GaussianDiffusion(object):
    def __init__(self, sigmas):
        self.sigmas = sigmas
        self.alphas = torch.sqrt(1 - sigmas**2)
        self.num_timesteps = len(sigmas)

    def add_noise(self, x0, noise, t):
        noise = torch.randn_like(x0) if noise is None else noise
        xt = _i(self.alphas, t, x0) * x0 + \
             _i(self.sigmas, t, x0) * noise
        return xt
    
    def get_velocity(self, x0, noise, t):
        v = _i(self.alphas, t, noise) * noise - _i(self.sigmas, t, x0) * x0
        return v
   ```
the loss is calculate with the following step:
```
noise = torch.randn_like(z_0)
prompt_embeddings = text_encoder(caption).detach()
z_t = noise_scheduler.add_noise(z_0, timesteps, noise)
target = noise_scheduler.get_velocity(z_0, noise, timesteps)
z_0_pred = unet(z_t, timesteps, prompt_embeddings)
loss = torch.nn.functional.mse_loss(z_0_pred.float(), target.float(), reduction=reduction)
```
Finetune related: @tastelikefeet  @Jintao-Huang

**Comment by oReano** [comment]
The training process seems ok. I suggest to check the calculation of sigmas and x_0.
Here are the codes for generating cosine beta schedules and calculating sigmas from betas:
```
def betas_to_sigmas(betas):
    return torch.sqrt(1 - torch.cumprod(1 - betas, dim=0))

def sigmas_to_betas(sigmas):
    square_alphas = 1 - sigmas ** 2
    betas = 1 - torch.cat([square_alphas[:1], square_alphas[1:] / square_alphas[:-1]])
    return betas

def cosine_schedule(n=1000, cosine_s=0.008):
    ramp = torch.linspace(0, 1, n + 1)
    square_alphas = torch.cos((ramp + cosine_s) / (1 + cosine_s) * torch.pi / 2) ** 2
    betas = (1 - square_alphas[1:] / square_alphas[:-1]).clamp(max=0.999)
    return betas_to_sigmas(betas)
```

Also, we use v-prediction loss function, so x_0 is calculated by:
`x0 = alphas * xt - sigmas * out`

**Comment by hpoghos** [comment]
[oReano](https://github.com/oReano) thanks for the answer,
you shared the `sigmas_to_betas` function, could you elaborate where it is used? As I understand we already have the sigmas why go back to betas?

**Comment by oReano** [comment]
Just in case you need it   haha :)

**Comment by zighem25** [comment]
@hpoghos  Which kind of data you are using to finetune the V2V model?

**Comment by hpoghos** [comment]
@zighem25 I'm using an upscaled WebVid-10M subset

**Comment by rob-hen** [comment]
Hi @oReano, you are doing an awesome job. Thank you so much for the help!

I also wanted to fine-tune the model and confirm the issues mentioned by @hpoghos .

Can it be that the noise scheduler should be:
```python
def cosine_scheduler(n=1000, cosine_s=0.008):
    ramp = torch.linspace(0, 1, n + 1)
    square_alphas = torch.cos(
        (ramp + cosine_s) / (1 + cosine_s) * torch.pi / 2) ** 2
    square_alphas = square_alphas / \
        math.cos((0+cosine_s) / (1+cosine_s) * math.pi / 2) ** 2
    betas = (1 - square_alphas[1:] / square_alphas[:-1])  # .clamp(max=0.999)

    return betas_to_sigmas(betas)
``` 

So first of all, I removed the clamp function. Since `zero_terminal_snr` is used during denoising, training should also allow 0 SNR.   Also, normally we [define](https://arxiv.org/pdf/2305.08891.pdf) the cosine schedule as 
$\min(1 - \frac{\overline{\alpha_{t}}}{\overline{\alpha_{t-1}}},0.999)$ , with $\overline{\alpha_{t}} = \frac{f(t)}{f(0)}$ and $f(t) = \cos(\frac{i+0.008}{1+0.008} \frac{\pi}{2})^2$ and $i = \frac{t-1}{T-1}$ . However, the normalisation $f(0)$ was missing in your code snippet. 


I am using the loss function 
```math code
||v - out ||_{2}^{2}
``` 
with 
```math code
v = \sqrt{\overline{\alpha_{t}}}\epsilon - \sqrt{1- \overline{\alpha_{t}}} z_{0},
``` 
and for the Video-To-Video model `unet`, input latent `z_0`, I obtain `out` via 


```python
timestep = torch.randint(0, 1000, (1,), device=device)
noise = torch.randn_like(z_0)
z_t = diffusion.add_noise(z_0,noise, timestep) # sqrt(\overline{alpha}_t)x_0+ sqrt(1-\overline{alpha}_t) * eps
out = unet(z_t,timestep,prompt_embedding)
``` 

Any idea what is missing?
